File: services/assesment-service/src/main/java/sgptt/assessvc/assesmentservice/security/RoleCheckAspect.java
package sgptt.assessvc.assesmentservice.security;

import jakarta.servlet.http.HttpServletRequest;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestClient;
import sgptt.assessvc.assesmentservice.dto.AuthDTO;

@Aspect
@Component
public class RoleCheckAspect {
    @Value("${env.data.auth.url}")
    private String authServiceUrl;

    private final RestClient.Builder restClientBuilder;
    private final HttpServletRequest request;

    @Autowired
    public RoleCheckAspect(RestClient.Builder restClientBuilder, HttpServletRequest request) {
        this.restClientBuilder = restClientBuilder;
        this.request = request;
    }

    @Before("@annotation(requiresRole)")
    public void checkRole(RequiresRole requiresRole) throws SecurityException {
        String token = request.getHeader("Authorization");

        if (token == null || token.isEmpty()) {
            throw new SecurityException("Token not found");
        }

        ResponseEntity<AuthDTO> response;

        RestClient request = restClientBuilder.baseUrl(authServiceUrl).build();

        try {
            response = request.get().uri("/api/auth/authorize/me").headers(headers -> {
                headers.add("Authorization", token);
            }).retrieve().toEntity(AuthDTO.class);
        } catch (HttpClientErrorException e) {
            if (e.getStatusCode() == HttpStatus.UNAUTHORIZED) {
                throw new SecurityException("Unauthorized: Invalid token");
            } else {
                throw new SecurityException("Error during authorization");
            }
        } catch (Exception e) {
            throw new SecurityException("Error during authorization");
        }

        if (response.getBody() == null) {
            throw new SecurityException("Invalid token");
        }

        String[] roles = requiresRole.value();

        for (String requiredRole : roles) {
            if (response.getBody().getRole().equals(requiredRole)) {
                return;
            }
        }

        throw new SecurityException("Unauthorized");
    }
}

File: services/protocols-service/src/main/kotlin/com/sgptt/protocolsservice/security/RoleCheckAspect.kt
package com.sgptt.protocolsservice.security

import com.sgptt.protocolsservice.model.dto.AuthDTO
import jakarta.servlet.http.HttpServletRequest
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.RestClient

@Aspect
@Component
class RoleCheckAspect @Autowired constructor(
	private val restClientBuilder: RestClient.Builder,
	private val request: HttpServletRequest,
) {
	@Value("\${env.data.auth.url}")
	private lateinit var authServiceUrl: String
	
	@Before("@annotation(requiresRole)")
	@Throws(SecurityException::class)
	fun checkRole(requiresRole: RequiresRole) {
		val token = request.getHeader("Authorization")
		if (token == null || token.isEmpty()) {
			throw SecurityException("Token not found")
		}
		val response: ResponseEntity<AuthDTO?>
		val request = restClientBuilder.baseUrl(authServiceUrl).build()
		try {
			response = request.get().uri("/api/auth/authorize/me").headers { headers: HttpHeaders ->
				headers.add(
					"Authorization",
					token
				)
			}.retrieve().toEntity(AuthDTO::class.java)
		} catch (e: HttpClientErrorException) {
			if (e.statusCode === HttpStatus.UNAUTHORIZED) {
				throw SecurityException("Unauthorized: Invalid token")
			} else {
				throw SecurityException("Error during authorization")
			}
		} catch (e: Exception) {
			throw SecurityException("Error during authorization")
		}
		
		if (response.body == null) {
			throw SecurityException("Invalid token")
		}
		
		val roles = requiresRole.roles
		
		for (requiredRole in roles) {
			if (response.body!!.role == requiredRole)
				return
		}
		
		throw SecurityException("Unauthorized")
	}
}